### ISSUE TITLE
clock remains in "unsynced" state, regardless of all-valid, or some-invalid peers

### ISSUE THREAD

**Issue by ghost** [openntpd-portable/openntpd-portable#25]
I've built/installed openntpd from git

```
cd /usr/local/src/openntpd
git log | head
    commit 827fa346509332556f18bd45f7cb4854f2736e0b
    Author: Brent Cook <bcook@openbsd.org>
    Date:   Tue Mar 29 13:54:32 2016 -0500

        More changelog updates

    commit 9182c78287774e8030df7fa1c1482c628f77dabb
    Author: Brent Cook <bcook@openbsd.org>
    Date:   Mon Mar 28 12:50:04 2016 -0500
```

Exec shows a number of peers some valid, some not -- for reasons that are not clear to me,

```
tail -f /var/log/ntpd.log 
    Apr 18 06:43:00 fwr ntpd[2660]: peer 45.56.95.44 now valid
    Apr 18 06:43:01 fwr ntpd[2660]: peer 216.218.254.202 now valid
    Apr 18 06:43:01 fwr ntpd[2660]: peer 128.138.140.44 now valid
    Apr 18 06:43:02 fwr ntpd[2660]: peer 204.123.2.5 now valid
    Apr 18 06:43:02 fwr ntpd[2660]: peer 69.75.229.42 now valid
    Apr 18 06:43:02 fwr ntpd[2660]: peer 69.36.224.15 now valid
    Apr 18 06:43:02 fwr ntpd[2660]: peer 45.56.95.44 now valid
    Apr 18 06:43:04 fwr ntpd[2660]: peer 67.18.187.111 now valid
    Apr 18 06:43:22 fwr ntpd[2660]: reply from 204.123.2.5: not synced (alarm), next query 3224s
    Apr 18 06:43:28 fwr ntpd[2660]: peer 216.218.254.202 now invalid
    Apr 18 06:48:17 fwr ntpd[2660]: peer 2001:4f8:0:2::1d now valid
    Apr 18 06:48:40 fwr ntpd[2660]: peer 2600:3c03:e000:17f::5 now valid
```

After 20+ mins, several peers remain invalid

```
ntpctl -sa | grep "not valid" -B1
    216.218.192.202 clock.fmt.he.net
        1  2  -  227s  300s             ---- peer not valid ----
    --
    216.218.254.202 clock.sjc.he.net
        1  2  -  230s  300s             ---- peer not valid ----
    --
    128.138.140.44 utcnist.colorado.edu
        1  2  1   84s  300s             ---- peer not valid ----
```

and the clock is still reporting as "unsynced"

```
Every 2.0s: ntpctl -s status                Mon Apr 18 06:49:09 2016

10/13 peers valid, clock unsynced
```

I note, watching over ~1hr, that, seemingly randomly, peers toggle between valid & invalid, and back again.

The openntpd server is serving as time source to my LAN clients.  The clients are, for the moment, running ntpd 4.2.8p6

The clients' status shows a preferred & active connect to the openntpd server, but appears to be 'stuck' using the refid of the one source that openntpd reports as "not synced (alarm)",

```
Every 2.0s: ntpq -np4                                                Mon Apr 18 06:46:10 2016

     remote           refid      st t when poll reach   delay   offset  jitter
==============================================================================
*10.15.3.17   204.123.2.5      2 u   18   64  370    0.345  -57.582  49.021
```

Note also the large offset/jitter.

Atm, about 2hrs after launching this openntpd instance

(1) the clock is still 'unsynced'
(2) peers keep dropping in/out of valid state
(3) the lan clients remain 'locked' to the same, single refid

First, what needs to be configured so that the openntpd returns, eventually a timely 'synced' state?

**Comment by dmosberger** [comment]
I'm seeing the same issue with the latest release (5.9p1) on Linux 4.4.0.  The problem seems to be coming from the `time_maxerror` variable: it gets initialized by the kernel to `NTP_PHASE_LIMIT` and never gets written by ntpd.  During each second overflow, the kernel then marks the clock as UNSYNCed because of this code:

```
    /* Bump the maxerror field */
    time_maxerror += MAXFREQ / NSEC_PER_USEC;
    if (time_maxerror > NTP_PHASE_LIMIT) {
        time_maxerror = NTP_PHASE_LIMIT;
        time_status |= STA_UNSYNC;
    }
```

I tried this hack:

```
$ diff -u compat/adjfreq_linux.c{~,}
--- compat/adjfreq_linux.c~ 2015-09-25 23:07:54.000000000 -0600
+++ compat/adjfreq_linux.c  2016-05-16 11:25:41.628333363 -0600
@@ -68,7 +68,10 @@
    struct timex txc = { 0 };

    txc.modes = MOD_STATUS;
-   if (!synced)
+   if (synced) {
+       txc.modes |= MOD_MAXERROR;
+       txc.maxerror = 0;
+   } else
        txc.status = STA_UNSYNC;
    if (adjtimex(&txc) == -1)
        log_warn("ntp_adjtime (3) failed");
```

and it fixes the problem:

```
$ adjtimex 
    mode:         0
-o  offset:       0 us
-f  freq.adjust:  1429078 (65536 = 1ppm)
    maxerror:     30500
    esterror:     16000000
    status:       0 ()
-p  timeconstant: 2
    precision:    1 us
    tolerance:    32768000
-t  tick:         10000 us
    time.tv_sec:  1463419786
    time.tv_usec: 12685
    return value: 0 (clock synchronized)
```

I'm not very familiar with the time-keeping code and I'm certain just setting `time_maxerror` to zero isn't the right thing to do here, but hopefully someone here knows what the correct fix would be.

**Comment by busterb** [comment]
This does not seem like a terrible approach. Maybe we should set STA_FREQHOLD as well in the synced case. This has been a long running problem on Solaris as well, I'm interested to see if this has the same effect there.

**Comment by busterb** [comment]
I added the patch above for testing. @pbhenson if you can test, how do things look in your Solaris environment? I found another oddball constant when examining the Chrony source in this area.

**Comment by pbhenson** [comment]
Sorry I didn't get to look at this sooner, I've been swamped at the day job :(. I see you've released 6.0p1 with this fix in it, I've got that built and running on Gentoo, RHEL6, Solaris 10, and OmniOS. I've never really had I problem with invalid peers, but my Solaris and OmniOS boxes still flap back and forth between sync'd and unsync'd. I'll let you know if that changes under 6.0p1, thanks.

**Comment by pacocc** [comment]
I'm having the same problem on a multi-cpu FreeBSD 11.0 system running 6.0p1 (apparently using with the above hack/patch).  Clock can be accurate to <=1s pre-start, using "-s" or not, but peers are always listed as "not valid".  Debugging indicates "no reply from ... received in time" despite tcpdump logging timely packets.  FWIW, the identical configuration worked on the previous hardware which had a single cpu.

**Comment by ghost** [comment]
I'm no longer using openntpd; switched to chrony on all boxes.  No more sync problems ...

Closing for my needs.

**Comment by pbhenson** [comment]
Sorry for the delay in response. I still haven't seen this problem on any of my Linux systems, this change hasn't made any real difference on them. We decommissioned our last Solaris 10 systems, but before we did they were still flapping sync, and our illumos/omnios systems still are. However, they never say the peers are invalid, they just flip between saying the clock is synced/not synced, so I don't know that this had anything really to do with it.
